corehq/motech/repeaters/tasks.py

```diff
@@ -143,39 +144,57 @@ def retry_process_repeat_record(repeat_record_id, domain):
-        repeat_record.save()
-        return
+        if not domain_can_forward(repeat_record.domain) or repeat_record.exceeded_max_retries:
+            # When creating repeat records, we check if a domain can forward so
+            # we should never have a repeat record associated with a domain that
+            # cannot forward, but this is just to be sure
+            repeat_record.cancel()
+            repeat_record.save()
+            return
 
-    if repeat_record.repeater.is_deleted:
-        repeat_record.cancel()
-        repeat_record.save()
-        return
+        if repeat_record.repeater.is_deleted:
+            repeat_record.cancel()
+            repeat_record.save()
+            return
 
-    try:
-        if repeat_record.repeater.is_paused or toggles.PAUSE_DATA_FORWARDING.enabled(repeat_record.domain):
-            # postpone repeat record by MAX_RETRY_WAIT so that it is not fetched
-            # in the next check to process repeat records, which helps to avoid
-            # clogging the queue
-            repeat_record.postpone_by(MAX_RETRY_WAIT)
-        elif rate_limit_repeater(repeat_record.domain):
-            # Spread retries evenly over the range defined by RATE_LIMITER_DELAY_RANGE
-            # with the intent of avoiding clumping and spreading load
-            repeat_record.postpone_by(random.uniform(*RATE_LIMITER_DELAY_RANGE))
-        elif repeat_record.is_queued():
-            with TimingContext() as timer:
-                repeat_record.fire()
-            # round up to the nearest millisecond, meaning always at least 1ms
-            report_repeater_usage(repeat_record.domain, milliseconds=int(timer.duration * 1000) + 1)
-    except Exception:
-        logging.exception('Failed to process repeat record: {}'.format(repeat_record.id))
+        try:
+            if repeat_record.repeater.is_paused or toggles.PAUSE_DATA_FORWARDING.enabled(repeat_record.domain):
+                # postpone repeat record by MAX_RETRY_WAIT so that it is not fetched
+                # in the next check to process repeat records, which helps to avoid
+                # clogging the queue
+                repeat_record.postpone_by(MAX_RETRY_WAIT)
+                action = 'paused'
+            elif rate_limit_repeater(repeat_record.domain):
+                # Spread retries evenly over the range defined by RATE_LIMITER_DELAY_RANGE
+                # with the intent of avoiding clumping and spreading load
+                repeat_record.postpone_by(random.uniform(*RATE_LIMITER_DELAY_RANGE))
+                action = 'rate_limited'
+            elif repeat_record.is_queued():
+                with TimingContext() as fire_timer:
+                    repeat_record.fire()
+                # round up to the nearest millisecond, meaning always at least 1ms
+                report_repeater_usage(repeat_record.domain, milliseconds=int(fire_timer.duration * 1000) + 1)
+                time_spent_waiting = fire_timer.subs[0].duration
```

Comment:
is there a way to refer to the particular sub-timer by name? Hard for me to tell if this refers to the same thing as fire_timer.duration, or refers to the `endpoint_timing` above

Reply:
yeah I didn't love this. In hindsight, I can convert it to a dictionary and navigate the subs field that way.

Reply:
Eh I didn't think that through fully. Ended up using a list comprehension inbaea03a96b2b5882e575ada962b506b6d0b8d291. Still doesn't feel great but short of adding a method to the TimingContext, I think it is the best option.